### DISCUSSION TITLE
Generalize NLOHMANN_DEFINE_TYPE

### DISCUSSION THREAD

**Discussion by kunaltyagi** [nlohmann/json#2828] [Ideas]
Right now, NLOHMANN_DEFINE_TYPE macro allows only `json` aka `basic_json<>` conversions. Would it make sense to:
1. Define the functions on templated `basic_json<T>` or perhaps `JSON` with some constraints
2. Specialize for `ordered_json` because it'd be faster to reserve + emplace_back in `to_json` and iterate using index + compare key + copy.

I can create a PR if this makes sense.

**Comment by nlohmann** [comment]
Part of this has already landed: `NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE` and friends are now templated over any `BasicJsonType` (via `NLOHMANN_JSON_BASIC_TYPE_TEMPLATE`), so the macros already work with `ordered_json` and other custom `basic_json` instantiations, addressing point 1. The second half of the proposal — a dedicated `ordered_json` fast path using reserve+emplace_back and index-based key comparison instead of the generic path — does not appear to be implemented and remains a legitimate, if micro-optimization-flavored, open idea.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
